ntoskrnl/include/internal/i386/mm.h

```diff
@@ -128,21 +117,68 @@
 /* On x86, these two are the same */
 #define MI_WRITE_VALID_PPE MI_WRITE_VALID_PTE
 
-/* Convert an address to a corresponding PTE */
-#define MiAddressToPte(x) \
-    ((PMMPTE)(((((ULONG)(x)) >> 12) << 2) + PTE_BASE))
+/*  Translating virtual addresses to physical addresses
+        (See: "Intel� 64 and IA-32 Architectures Software Developer�s Manual
+              Volume 3A: System Programming Guide, Part 1, CHAPTER 4 PAGING")
+    Page directory (PD) and Page table (PT) definitions
+    Page directory entry (PDE) and Page table entry (PTE) definitions
+*/
+
+/* Maximum number of page directories pages */
+#ifndef _X86PAE_
+#define PD_COUNT 1        /* Only one page directory page */
+#else
+#define PD_COUNT (1 << 2) /* The two most significant bits in the VA */
+#endif
+
+/* The number of PTEs on one page of the PT */
+#define PTE_PER_PAGE (PAGE_SIZE / sizeof(MMPTE))
+
+/* The number of PDEs on one page of the PD */
+#define PDE_PER_PAGE (PAGE_SIZE / sizeof(MMPDE))
+
+/* Maximum number of PDEs */
+#define PDE_PER_SYSTEM (PD_COUNT * PDE_PER_PAGE)
+
+/* TODO: It seems this variable is not needed for x86 */
+#define PPE_PER_PAGE 1
+
+/* Maximum number of pages for 4 GB of virtual space */
+#define MI_MAX_PAGES ((1ull << 32) / PAGE_SIZE)
+
+/* Base addresses for page tables */
+#define PTE_BASE (ULONG_PTR)0xC0000000
+#define PTE_TOP  (ULONG_PTR)(PTE_BASE + (MI_MAX_PAGES * sizeof(MMPTE)) - 1)
+#define PTE_MASK (PTE_TOP - PTE_BASE)
+
+#define MI_SYSTEM_PTE_BASE (PVOID)MiAddressToPte(NULL)
+
+/* Base addreses for page directories */
+#define PDE_BASE (ULONG_PTR)MiAddressToPte(PTE_BASE)
+#define PDE_TOP  (ULONG_PTR)(PDE_BASE + (PDE_PER_SYSTEM * sizeof(MMPDE)) - 1)
+#define PDE_MASK (PDE_TOP - PDE_BASE)
+
+/* The size of the virtual memory area that is mapped using a single PDE */
+#ifndef PDE_MAPPED_VA
```

Comment:
Where else would it be defined ?

Reply:
https://github.com/reactos/reactos/blob/d37799ea3e52a74493eae40f861b4a5f6a3b8b8f/ntoskrnl/mm/ARM3/miarm.h#L22
Without this ifndef, hal failed compilation.

Reply:
Incidentally there also exists a `PDE_PER_PAGE`  in our code base.

Reply:
But it wasn’t used before in ntoskrnl/include/internal/i386/mm.h ...

Reply:
Maybe because someone made a copy-paste error and forgot to replace where needed, PTE by PDE?

Reply:
There is no error, I just added PDE_MAPPED_VA in macros that are used by HAL.
https://git.reactos.org/?p=reactos.git;a=blob;f=hal/halx86/include/halp.h;hb=cefc378b0d64a30d246f0ff7a8f71e97de337f78#l530

Reply:
Then maybe remove it from miarm.h while we're at it ? At some point, this header will have to disappear and be merged with the other ones.